### PR TITLE
Added SimProcs for linux::dup{1,2,3}

### PR DESCRIPTION
**Pull request by Alexeyan**
Hi, 
i made simprocs for dup, dup2 and dup3 syscalls. 
Open for change requests.

Best regards,
Alexeyan

### CODE REVIEW

**Comment by Alexeyan** [comment]
So 1, 2 and 3 are done. 
But i double checked and from my tests it seems like the positions aren't independent.
It looks like it's just a different number, but refering to the same stream.

File: angr/procedures/posix/dup.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class dup(angr.SimProcedure):  #pylint:disable=W0622
+    def run(self, oldfd): # pylint:disable=arguments-differ
+        if oldfd not in self.state.posix.fd.keys():
+            return self.state.libc.ret_errno('EBADF')
+
+        # The new fd gets the lowest free number, so we search
+        newfd = len(self.state.posix.fd.keys()) # e.g. '3' for [0, 1, 2]
+        for i, fd in enumerate(self.state.posix.fd.keys()):
```
Comment (rhelmot) [inline angr/procedures/posix/dup.py]: this needs to be enumerate(sorted(keys)) in order for this algorithm to work. tbh I would just scan upwards from 0 and pick the first free one.

**Comment by rhelmot** [comment]
ok, um, so my entire worldview for how linux architected file descriptors and the reason I designed simfile the way I did just crumbled beneath my fingertips. I will need some time to process this and then we will need to refactor everything, again

fix the above issue and I'll hit merge

File: angr/procedures/posix/dup.py (hunk 1 above)
Comment (Alexeyan) [reply angr/procedures/posix/dup.py]: Done.

**Comment by rhelmot** [comment]
you, um, didn't match the parentheses in that commit

also you should make sure there's a newline at the end of the file or it'll fail lint

**Comment by rhelmot** [comment]
thanks!

**Pull request merged**